### ISSUE TITLE
`hook-use-state` rule doesn't allow omitting unused setter

### ISSUE THREAD

**Issue by drbr** [jsx-eslint/eslint-plugin-react#3268]
Thanks for implementing the new `hook-use-state` rule! I'm excited to use it on my team's codebase to enforce consistently-named `[state, setState]` pairs.

Occasionally we have code like this:
```
const [ constantValue ] = useState(() => expensivelyComputeConstantValue());
```

We use `useState` for this because it's the most straightforward way to guarantee that the initializer is run only once. While there may be more idiomatic ways to do this depending on the use case, I don't see a practical downside to using `useState` this way, so I think would be helpful if the hook allowed this pattern. What do you think about changing `hook-use-state` so that a one-argument array is considered correct?

**Comment by ljharb** [comment]
The proper thing to do here is `useMemo`, not `useState`.

**Comment by drbr** [comment]
As we understand it, useMemo currently runs the function only once, but doesn't guarantee that it'll always behave that way in the future. Hence we're using useState instead.

**Comment by ljharb** [comment]
Of course it does? It's based on the provided dependency array.

**Comment by ljharb** [comment]
If the value depends on the props, then react *should* be free to recalculate it whenever it wants.

if it doesn't, then why is it computed inside the component? Memoize it at module level, with your own memoization helper.

**Comment by ljharb** [comment]
Why is computing a unique ID expensive? That's definitely the kind of thing useMemo is meant for.

> would it be in character for the React lint plugin to suggest switching from useState to useMemo if it detects that the state setter isn't being used?

That definitely shouldn't be an autofix, but it's a reasonable suggestion for the rule to make, sure!

**Comment by silverwind** [comment]
I have to raise a pattern I use sometimes, for incomplete code that is to be completed later:

```js
const [thing, _setThing] = useState(null);
```

It violates `react/hook-use-state` because of the variable naming, but I think an option to allow leading underscore on either of the variable names might be nice to have.

**Comment by ljharb** [comment]
Seems like it'd be more obviously a TODO to add an eslint override comment instead?

**Comment by silverwind** [comment]
I still feel the rule could accept a `ignorePattern` like eslint core rule `no-unused-vars` takes as it's essentially the same, a unused var. I configure that eslint rule like this:

```yaml
no-unused-vars: [2, {args: all, argsIgnorePattern: ^_, varsIgnorePattern: ^_, caughtErrorsIgnorePattern: ^_, destructuredArrayIgnorePattern: ^_, ignoreRestSiblings: false}]
```

So for consistency sake with that config, I would suggest :

```yaml
  react/hook-use-state: [2, {ignorePattern: ^_}]
```

**Comment by ljharb** [comment]
Having a rule take a regex string is a very dangerous path that results in a ton of CVEs, so I wouldn't want to replicate that.

**Comment by silverwind** [comment]
If you are afraid of regex runtime, you can use https://github.com/sindresorhus/super-regex, but generally I think eslint configs are unlikely to receive untrusted inputs.

**Comment by ljharb** [comment]
I agree, but they are highly likely to get CVEs about untrusted inputs, which cause a lot of churn and cost.

That package (along with all of that author's packages) aren't viable options, since they're ESM-only.
